tracpro/contacts/utils.py

```diff
@@ -0,0 +1,271 @@
+    :param last_time: the last time we pulled contacts, if None, sync all contacts
+    :param delete_blocked: if True, delete the blocked contacts
+    :return: tuple containing list of UUIDs for created, updated, deleted and failed contacts
+    """
+    # get all remote contacts
+    client = get_client(org)
+    updated_incoming_contacts = []
+    if last_time:
+        updated_incoming_contacts = client.get_contacts(after=last_time)
+    else:
+        updated_incoming_contacts = client.get_contacts()
+
+    # get all existing contacts and organize by their UUID
+    existing_contacts = contact_class.objects.filter(org=org)
+    existing_by_uuid = {contact.uuid: contact for contact in existing_contacts}
+
+    created_uuids = []
+    updated_uuids = []
+    deleted_uuids = []
+    failed_uuids = []
+
+    for updated_incoming in updated_incoming_contacts:
+        # delete blocked contacts if deleted_blocked=True
+        if updated_incoming.blocked and delete_blocked:
+            deleted_uuids.append(updated_incoming.uuid)
+
+        elif updated_incoming.uuid in existing_by_uuid:
+            existing = existing_by_uuid[updated_incoming.uuid]
+
+            diff = temba_compare_contacts(updated_incoming, existing.as_temba(), fields, groups)
+
+            if diff or not existing.is_active:
+                try:
+                    kwargs = contact_class.kwargs_from_temba(org, updated_incoming)
+                except ValueError:
+                    failed_uuids.append(updated_incoming.uuid)
+                    continue
+
+                for field, value in six.iteritems(kwargs):
+                    setattr(existing, field, value)
+
+                existing.is_active = True
+                existing.save()
+
+                updated_uuids.append(updated_incoming.uuid)
+        else:
+            try:
+                kwargs = contact_class.kwargs_from_temba(org, updated_incoming)
+            except ValueError:
+                failed_uuids.append(updated_incoming.uuid)
+                continue
+
+            contact_class.objects.create(**kwargs)
+            created_uuids.append(kwargs['uuid'])
+
+    # any contact that has been deleted from rapidpro
+    # should also be deleted from dash
+    # if last_time was passed in, just get contacts deleted after the last time we synced
+    if last_time:
+        deleted_incoming_contacts = client.get_contacts(deleted=True, after=last_time)
+    else:
+        deleted_incoming_contacts = client.get_contacts(deleted=True)
+    for deleted_incoming in deleted_incoming_contacts:
+        deleted_uuids.append(deleted_incoming.uuid)
+    existing_contacts.filter(uuid__in=deleted_uuids).update(is_active=False)
+
+    return created_uuids, updated_uuids, deleted_uuids, failed_uuids
+
+
+def temba_compare_contacts(first, second, fields=None, groups=None):
+    """
+    Compares two Temba contacts to determine if there are differences. Returns
+    first difference found.
```

Comment:
It seems like this is intended to compare two versions of the same Temba contact, so maybe the docstring could say something like that.  Also, maybe add a brief description of the `fields` and `groups` arguments' purpose?